src/facade/reply_builder.cc

```diff
@@ -228,34 +249,54 @@ void SinkReplyBuilder2::WritePiece(std::string_view str) {
 
 void SinkReplyBuilder2::WriteRef(std::string_view str) {
   NextVec(str);
+  ref_indices_.push_back(vecs_.size() - 1);
   total_size_ += str.size();
 }
 
 void SinkReplyBuilder2::Flush() {
-  auto ec = sink_->Write(vecs_.data(), vecs_.size());
-  if (ec)
+  auto& reply_stats = tl_facade_stats->reply_stats;
+
+  send_active_ = true;
+  uint64_t before_ns = util::fb2::ProactorBase::GetMonotonicTimeNs();
+  reply_stats.io_write_cnt++;
+  reply_stats.io_write_bytes += total_size_;
+
+  if (auto ec = sink_->Write(vecs_.data(), vecs_.size()); ec)
     ec_ = ec;
 
+  uint64_t after_ns = util::fb2::ProactorBase::GetMonotonicTimeNs();
+  reply_stats.send_stats.count++;
+  reply_stats.send_stats.total_duration += (after_ns - before_ns) / 1'000;
+  send_active_ = false;
+
+  if (buffer_.InputLen() * 2 > buffer_.Capacity())  // If needed, grow backing buffer
+    buffer_.Reserve(std::min(kMaxBufferSize, buffer_.Capacity() * 2));
+
+  total_size_ = 0;
   buffer_.Clear();
   vecs_.clear();
-  total_size_ = 0;
+  ref_indices_.clear();
 }
 
 void SinkReplyBuilder2::FinishScope() {
-  // If batching or aggregations are not enabled, flush
-  Flush();
-
-  // TODO: otherwise iterate over vec_ and copy items to buffer_
-  // whilst also updating their pointers
-}
-
-bool SinkReplyBuilder2::IsInBuf(const void* ptr) const {
-  auto ib = buffer_.InputBuffer();
-  return ptr >= ib.data() && ptr <= ib.data() + ib.size();
+  if (!batched_ || total_size_ * 2 >= kMaxBufferSize)
+    return Flush();
+
+  size_t required = total_size_ - buffer_.InputLen();
+  if (required > buffer_.AppendLen())
+    return Flush();
+
+  // Copy all extenral references to buffer to safely keep batching
+  for (unsigned i : ref_indices_) {
```

Comment:
Need to tune this properly. The idea is that if we have lots of external refs stashed close to the limit, there's no point in copying them all if we can finish our batch earlier

Reply:
extenral -> external

Reply:
I did not understand your comment btw

Reply:
Suppose your batching responses and your current state is

```
{
$100500
*long string as extern ref*
\r\n
... 10 more like this ...
}
```

And now your scope ends, meaning that all your string refs become invalid, but total size is still less than the max batch limit. If after copying you end up with ~90% of the limit, it would be much more efficient to just send an incomplete batch instead of copying all the data

Reply:
Yes